pkg/segment/writer/segstore.go

```diff
@@ -1034,56 +1079,37 @@ func (ss *SegStore) flushBloomIndex(cname string, bi *BloomIndex) uint64 {
 
 	fname := fmt.Sprintf("%s_%v.cmi", ss.SegmentKey, xxhash.Sum64String(cname))
 
-	bffd, err := os.OpenFile(fname, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
+	bytesWritten, err := ss.writeBloomToFile(fname, bi)
 	if err != nil {
-		log.Errorf("flushBloomIndex: open failed fname=%v, err=%v", fname, err)
-		return 0
-	}
-
-	defer bffd.Close()
-
-	var buf bytes.Buffer
-	bufWriter := bufio.NewWriter(&buf)
-
-	// there is no accurate way to find exactly how many bytes the write.to is going to write
-	// and we need that number , so that we write it first and then the actual bloom data
-	// hence this messiness to write it to some buffer, get the bytesWritten count and then do
-	// the actual write
-	bytesWritten, bferr := bi.Bf.WriteTo(bufWriter)
-	if bferr != nil {
-		log.Errorf("flushBloomIndex: write buf failed fname=%v, err=%v", fname, bferr)
 		return 0
 	}
 
-	bytesWritten += utils.LEN_BLKNUM_CMI_SIZE // for blkNum
-	bytesWritten += 1                         // reserver for CMI_Type
-
-	// copy the size of blockBloom in uint32
-	if _, err = bffd.Write(toputils.Uint32ToBytesLittleEndian(uint32(bytesWritten))); err != nil {
-		log.Errorf("flushBloomIndex: bloomsize write failed fname=%v, err=%v", fname, err)
+	// update the bloom size in the file
+	bffd, err := os.OpenFile(fname, os.O_WRONLY, 0644)
+	if err != nil {
+		log.Errorf("flushBloomIndex: open failed fname=%v for writing the bloom size, err=%v", fname, err)
 		return 0
 	}
+	defer bffd.Close()
 
-	// copy the blockNum
-	if _, err = bffd.Write(toputils.Uint16ToBytesLittleEndian(ss.numBlocks)); err != nil {
-		log.Errorf("flushBloomIndex: bloomsize write failed fname=%v, err=%v", fname, err)
+	_, err = bffd.Seek(-int64(bytesWritten), io.SeekEnd)
```

Comment:
I think this is fine, but there might be a concurrency issue. Is `flushBloomIndex()` ever called concurrently? If so, then seeking from the end like this could put us at the wrong location. If that's an issue, I believe seeking from the start should fix the issue (and in that case, you could use `WriteAt()` instead of seeking and then writing

Reply:
Addressed in the latest commit.